documentation/src/main/asciidoc/reference/introduction.adoc

```diff
@@ -977,26 +977,54 @@ sessionFactory.withTransaction( (session, tx) -> session.persist(book) )
 
 This is probably the most convenient thing to use most of the time.
 
-=== Vert.x context
+== Integrating with Vert.x
 
 :vertx-context-introduction: https://vertx.io/blog/an-introduction-to-the-vert-x-context-object/
-
-The session is not thread-safe, using it in different threads may cause issues that are hard to debug.
-When it's created, a session is bound to a {vertx-context-introduction}[Vert.x context]
-and needs to be used in the same context after that.
-
-Hibernate Reactive will take care of this when the application uses `withSession`,
-`withStatelessSession` or `withTransaction` (See <<_obtaining_a_reactive_session>>).
-
-When using `openSession` or `openStatelessSession`, one need to make sure
-to use it in the right Vert.x context or the following exception might occur:
+:vertx-local-context: https://vertx.io/docs/apidocs/io/vertx/core/Context.html#getLocal-java.lang.Object-
+:vertx-get-context: https://vertx.io/docs/apidocs/io/vertx/core/Vertx.html#getOrCreateContext--
+:vertx-runon-context: https://vertx.io/docs/apidocs/io/vertx/core/Context.html#runOnContext-io.vertx.core.Handler-
+
+At runtime, interaction with the database happens on a Vert.x thread,
+typically the event loop thread. When you write code that creates and
+destroys Hibernate Reactive sessions, it's important to understand how
+sessions relate to threads and {vertx-context-introduction}[Vert.x contexts].
+
+=== Sessions and Vert.x contexts
+
+TIP: Remember how in regular old Hibernate JPA, you're not supposed to
+share a session between multiple threads? Well, the idea here is essentially
+similar, it's just that the notion of a "thread" is a little more slippery,
+or at least more _technical_. You need to be able to replace the idea of a
+"thread" with the idea of a chain of callbacks occurring on a reactive stream,
+all within the scope of a certain Vert.x _local context_.
+
+When you create a session using `withSession()` or `withTransaction()`, it's
+automatically associated with the current Vert.x
+{vertx-local-context}[local context], and propagates with the local context,
+as mentioned above in <<_obtaining_a_reactive_session>>. And you're only
+allowed to use the session from the thread that owns this local context. If
+you screw up, and use it from a different thread, you might see this error:
 
 [source, JAVA, indent=0]
 ----
 HR000068: This method should exclusively be invoked from a Vert.x EventLoop thread; ...
 ----
 
-One way to run the code in the correct Vert.x context is the following:
+On the other hand, if you use `openSession()`, you'll have to manage the
+association between sessions and contexts yourself. Now, that's in principle
+straightforward, but you'd be surprised how often people mess up.
```

Comment:
I assure you, nobody is going to be surprised :-)

Reply:
hahaha

Reply:
If I was being completely honest, I would have written:

> Now, that seems pretty straightforward *to me*, and I'm continually surprised how often *you lot* mess it up.

Reply:
With an eyeroll.gif